charms/istio-pilot/lib/charms/istio_pilot/v0/ingress.py

```diff
@@ -0,0 +1,305 @@
+                to "/"
+            namespace: the namespace the service is in; default to the current model
+            per_unit_routes: whether or not to create URLs which map to specific units;
+                the URLs will have their own prefix of f"{prefix}-unit-{unit_num}" (with
+                tailing slashes handled appropriately)
+        """
+        super().__init__(charm, f"ingress-requirer-{relation_name}")
+        self.charm = charm
+        self.relation_name = relation_name
+
+        self._validate_relation_meta()
+
+        self.status = self._get_status()
+
+        self._request_args = {
+            "port": port,
+            "service": service,
+            "prefix": prefix,
+            "rewrite": rewrite,
+            "namespace": namespace,
+            "per_unit_routes": per_unit_routes,
+        }
+        if any(self._request_args.values()) and not self._request_args["port"]:
+            raise TypeError("Missing required argument: 'port'")
+
+        if charm.unit.is_leader():
+            self.framework.observe(charm.on[relation_name].relation_created, self._check_provider)
+            self.framework.observe(charm.on[relation_name].relation_changed, self._check_provider)
+            self.framework.observe(charm.on[relation_name].relation_broken, self._lost_provider)
+            self.framework.observe(charm.on.leader_elected, self._check_provider)
+
+    def _get_status(self):
+        if not self.charm.unit.is_leader():
+            return None
+        if not self.charm.model.relations[self.relation_name]:
+            # the key will always exist but may be an empty list
+            return BlockedStatus(f"Missing relation: {self.relation_name}")
+        try:
+            self._get_interface()
+        except NoCompatibleVersions:
+            return BlockedStatus(f"Relation version not compatible: {self.relation_name}")
+        except NoVersionsListed:
+            return WaitingStatus(f"Waiting on relation: {self.relation_name}")
+        else:
+            return None
+
+    def _get_interface(self):
+        """Get the SDI instance for the relation.
+
+        This provides defaults for the schema URL and supported versions so that
+        every client charm doesn't need to specify it, since they're already using
+        the versioned library which is inherently tied to a schema & version. It
+        still defers to what's in the metadata.yaml, however, so that local_sdi.py
+        can be used to inline the schema into the charm to avoid runtime network
+        access, if desired.
+        """
+        # Can't use self.charm.meta, unfortunately, because it doesn't preserve
+        # the schema or versions fields.
+        meta = yaml.safe_load(Path("metadata.yaml").read_text())
+        endpoint_spec = meta["requires"][self.relation_name]
+        schema = get_schema(endpoint_spec.get("schema", SCHEMA_URL))
+        versions = endpoint_spec.get("versions", SCHEMA_VERSIONS)
```

Comment:
The wrapper library's code is inherently tied to the specific schema versions it supports, just like the istio-pilot charm's code is inherently tied to the versions *it* supports. Specifying the version in the library is equivalent to the charm specifying the versions in it's `metadata.yaml` and doing it here keeps it closest to, and coupled with, the code that actually has that compatibility requirement. If we offload it to the consuming charm, then they can put whatever versions they want and the library will break.

The schema URL default, OTOH, is just convenience so that the consuming charm authors don't have to copypasta that when they're already copypastaing the entire library in via `charmcraft`.  Also, the intention is that pretty much everyone should add the `local_sdi` bit to their `charmcraft.yaml` so that it doesn't have external runtime dependencies, which is why it allows for the `metadata.yaml` to override it.